File: nim/day16.nim
import strutils, sequtils, algorithm


const
  instructions = readFile("./inputs/16.txt").split(',')
  iterations = 1_000_000_000
  programs = "abcdefghijklmnop"


type Parsed = object
  first, pa, pb: char
  second, xa, xb: int


proc parse(instruction: string): Parsed =
  let
    first = instruction[0]
    rest = instruction[1 .. instruction.high]
  var x: seq[int]

  case first
    of 's':
      result = Parsed(first: first, second: rest.parseInt)
    of 'x':
      x = rest.split('/').map(parseInt)
      result = Parsed(first: first, xa: x[0], xb: x[1])
    of 'p':
      result = Parsed(first: first, pa: rest[0], pb: rest[^1])
    else: discard


proc cleanUp(instructions: seq[string]): seq[Parsed] =
  result = @[]
  for instruction in instructions:
    result.add(parse(instruction))


let cleaned = instructions.cleanUp()


proc dance(dancers: string): string =
  result = dancers
  var a, b: int
  for element in cleaned:
    case element.first
    of 's':
      discard result.rotateLeft(-element.second)
    of 'x':
      swap(result[element.xa], result[element.xb])
    of 'p':
      a = result.find(element.pa)
      b = result.find(element.pb)
      result[a] = element.pb
      result[b] = element.pa
    else: discard


proc longDance(dancers: string): string =
  var
    dancers = dancers
    seen = @[dancers]
  for i in 1 .. iterations:
    dancers = dancers.dance()
    if dancers in seen:
      return seen[iterations mod i]
    seen.add(dancers)


echo programs.dance()
echo programs.longDance()

File: python/day16.py
from collections import deque


with open('./inputs/16.txt') as f:
    instructions = f.readline().split(',')


def clean_up(instructions):
    result = []
    for instr in instructions:
        first = instr[0]
        if first == 's':
            result.append((first, int(instr[1:])))
        elif first == 'x':
            a, b = map(int, instr[1:].split('/'))
            result.append((first, a, b))
        else:
            result.append((first, instr[1], instr[-1]))
    return result


def dance(dancers):
    dancers = deque(dancers)
    for first, *rest in cleaned:
        if first == 's':
            dancers.rotate(rest[0])
        elif first == 'x':
            a, b = rest
            dancers[a], dancers[b] = dancers[b], dancers[a]
        elif first == 'p':
            x, y = rest
            a = dancers.index(x)
            b = dancers.index(y)
            dancers[a], dancers[b] = y, x
    return ''.join(dancers)


def long_dance(dancers, iterations=1_000_000_000):
    seen = [dancers]
    for i in range(1, iterations):
        dancers = dance(dancers)
        if dancers == programs:
            return seen[iterations % i]
        seen.append(dancers)


programs = 'abcdefghijklmnop'
cleaned = clean_up(instructions)

print(dance(programs))
print(long_dance(programs))
